pkg/cluster/internal/create/actions/join/join.go

```diff
@@ -0,0 +1,267 @@
+	for _, node := range secondaryControlPlanes {
+		if err := runKubeadmJoinControlPlane(ctx, allNodes, &node); err != nil {
+			return err
+		}
+	}
+
+	ctx.Status.End(true)
+	return nil
+}
+
+func joinWorkers(
+	ctx *actions.ActionContext,
+	allNodes []nodes.Node,
+	workers []nodes.Node,
+) error {
+	ctx.Status.Start("Joining worker nodes to Kubernetes ☸")
+	defer ctx.Status.End(false)
+
+	// TODO(bentheelder): this should be concurrent
+	for _, node := range workers {
+		if err := runKubeadmJoin(ctx, allNodes, &node); err != nil {
+			return err
+		}
+	}
+
+	ctx.Status.End(true)
+	return nil
+}
+
+// runKubeadmJoinControlPlane executes kubadm join --control-plane command
+func runKubeadmJoinControlPlane(
+	ctx *actions.ActionContext,
+	allNodes []nodes.Node,
+	node *nodes.Node,
+) error {
+	// get the join address
+	joinAddress, err := getJoinAddress(ctx, allNodes)
+	if err != nil {
+		// TODO(bentheelder): logging here
+		return err
+	}
+
+	// creates the folder tree for pre-loading necessary cluster certificates
+	// on the joining node
+	if err := node.Command("mkdir", "-p", "/etc/kubernetes/pki/etcd").Run(); err != nil {
+		return errors.Wrap(err, "failed to join node with kubeadm")
+	}
+
+	// define the list of necessary cluster certificates
+	fileNames := []string{
+		"ca.crt", "ca.key",
+		"front-proxy-ca.crt", "front-proxy-ca.key",
+		"sa.pub", "sa.key",
+	}
+
+	// creates a temporary folder on the host that should acts as a transit area
+	// for moving necessary cluster certificates
+	tmpDir, err := fs.TempDir("", "")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	err = os.MkdirAll(filepath.Join(tmpDir, "/etcd"), os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	// get the handle for the bootstrap control plane node (the source for necessary cluster certificates)
+	controlPlaneHandle, err := nodes.BootstrapControlPlaneNode(allNodes)
+	if err != nil {
+		return err
+	}
+
+	// copies certificates from the bootstrap control plane node to the joining node
+	for _, fileName := range fileNames {
+		// sets the path of the certificate into a node
+		containerPath := path.Join("/etc/kubernetes/pki", fileName)
+		// set the path of the certificate into the tmp area on the host
+		tmpPath := filepath.Join(tmpDir, fileName)
+		// copies from bootstrap control plane node to tmp area
+		if err := controlPlaneHandle.CopyFrom(containerPath, tmpPath); err != nil {
+			return errors.Wrapf(err, "failed to copy certificate %s", fileName)
+		}
+		// copies from tmp area to joining node
+		if err := node.CopyTo(tmpPath, containerPath); err != nil {
+			return errors.Wrapf(err, "failed to copy certificate %s", fileName)
+		}
+	}
+
+	// run kubeadm join --control-plane
+	cmd := node.Command(
+		"kubeadm", "join",
+		// the join command uses the docker ip and a well know port that
+		// are accessible only inside the docker network
+		joinAddress,
+		// set the node to join as control-plane
+		"--experimental-control-plane",
```

Comment:
post 1.14, we should be able to replace the copy of certs (L126-170) with the new `--certificate-key` flag on `init / join`. this logic is now built in kubeadm.

Reply:
👍 

we'll still need this path for < 1.14 😬